presto-main/src/main/java/com/facebook/presto/operator/aggregation/histogram/GroupedTypedHistogram.java

```diff
@@ -0,0 +1,541 @@
+        }
+
+        throw new IllegalArgumentException(String.format("invalid enum %s", PROBE_TYPE));
+    }
+
+    //parallel arrays with data for
+    private void resizeNodeArrays(int newBucketCount)
+    {
+        // every per-bucket array needs to be updated
+        counts.ensureCapacity(newBucketCount);
+        valuePositions.ensureCapacity(newBucketCount);
+        nextPointers.ensureCapacity(newBucketCount);
+        valueHashList.ensureCapacity(newBucketCount);
+        groupIds.ensureCapacity(newBucketCount);
+    }
+
+    private long getHash(long groupId, long valueRawHash)
+    {
+        return (murmurHash3(valueRawHash) ^ murmurHash3(groupId));
+    }
+
+    private int getBucketIdForNode(int nodePointer, int mask)
+    {
+        long valueRawHash = valueHashList.get(nodePointer); // without mask
+        int bucketId = (int) (valueRawHash & mask);
+
+        return bucketId;
+    }
+
+    //short-lived abstraction that is basically a position into parallel arrays that we can treat as one data structure
+    private class ValueNode
+    {
+        // index into parallel arrays that are fields of a "node" in our hash table (eg counts, valuePositions)
+        private final int nodePointer;
+
+        /**
+         * @param nodePointer - index/pointer into parallel arrays of data structs
+         * @param isEmpty
+         */
+        ValueNode(int nodePointer)
+        {
+            checkState(nodePointer > -1, "ValueNode must point to a non-empty node");
+            this.nodePointer = nodePointer;
+        }
+
+        long getCount()
+        {
+            return counts.get(nodePointer);
+        }
+
+        int getValuePosition()
+        {
+            return valuePositions.get(nodePointer);
+        }
+
+        void add(long count)
+        {
+            counts.add(nodePointer, count);
+        }
+
+        /**
+         * given an output outputBlockBuilder, writes one row (key -> count) of our histogram
+         *
+         * @param valuesBlock - values.build() is called externally
+         */
+        void writeNodeAsRow(Block valuesBlock, BlockBuilder outputBlockBuilder)
+        {
+            type.appendTo(valuesBlock, getValuePosition(), outputBlockBuilder);
+            BIGINT.writeLong(outputBlockBuilder, getCount());
+        }
+    }
+
+    // short-lived class that wraps a position in int buckets[] to help handle treating
+    // it as a hash table with Nodes that have values
+    private class BucketDataNode
+    {
+        // index into parallel arrays that are fields of a "node" in our hash table (eg counts, valuePositions)
+        private final int bucketId;
+        private final ValueNode valueNode;
+        private final long valueHash;
+        private final int nodePointerToUse;
+        private final boolean isEmpty;
+
+        /**
+         * @param bucketId - index into the bucket array. It is writable. If ! EMPTY_BUCKET, it's value points to valid counts and pointers in
```

Comment:
What's "writable"? The bucketId field is final, so this is a little confusing.

Reply:
just that it's a mutable entity. perhaps it's like my open address hashing comment above and superfluous

it's writable either as
a. an empty entry
b. an existing entry

and caller doens't need to know which, since processEntry() handles cases
i'll remove since it's confusing

Reply:
done